sw/device/lib/dif/dif_sram_ctrl.h

```diff
@@ -11,256 +11,276 @@
- * @param[out] handle Out param for the initialized handle.
+ * @param base_addr The MMIO base address of the IP.
+ * @param[out] sram_ctrl Out param for the initialized handle.
  * @return The result of the operation.
  */
 OT_WARN_UNUSED_RESULT
-dif_sram_ctrl_result_t dif_sram_ctrl_init(dif_sram_ctrl_params_t params,
-                                          dif_sram_ctrl_t *handle);
+dif_result_t dif_sram_ctrl_init(mmio_region_t base_addr,
+                                dif_sram_ctrl_t *sram_ctrl);
 
 /**
- * Trigger one alert event.
+ * Performs SRAM scrambling.
+ *
+ * This is a compound operation covering both data and address "scrambling".
+ * In other words logical re-mapping of the physical addresses, followed by
+ * the entire SRAM overwriting with a pseudo-random data.
+ *
+ * This function should only be called when the data is no longer used.
+ * Additional implication is that the C runtime can be invalidated by the call
+ * to this function, and must be re-configured prior to any C code execution.
+ *
+ * The intention of this operation is to ensure that there is no predefined
+ * values or predictable data that could potentially make "unscrambling"
+ * easier.
+ *
+ * This operation is expected to take a significant amount of CPU cycles. The
+ * status can be checked via `dif_sram_ctrl_scramble_is_busy`, which is useful
+ * when a non-blocking work flow is desirable. Otherwise any SRAM access will
+ * automatically block until this operation has finished.
+ *
+ * The equivalent effect can be achieved by calling
+ * `dif_sram_ctrl_request_new_key`, followed by `dif_sram_ctrl_wipe`.
  *
  * @param sram_ctrl A SRAM Controller handle.
  * @return The result of the operation.
  */
 OT_WARN_UNUSED_RESULT
-dif_sram_ctrl_result_t dif_sram_ctrl_force_alert(
-    const dif_sram_ctrl_t *sram_ctrl);
+dif_result_t dif_sram_ctrl_scramble(const dif_sram_ctrl_t *sram_ctrl);
 
 /**
- * Lock execution enable register.
+ * Checks whether SRAM scramble is still progress.
  *
- * This function prevents any further writes to the
- * execution enable register. Future attempts to enable or disable
- * execution from SRAM will fail.
+ * Useful when a non-blocking work flow is desirable.
  *
  * @param sram_ctrl A SRAM Controller handle.
+ * @param[out] is_busy Out-param for the busy state.
  * @return The result of the operation.
  */
 OT_WARN_UNUSED_RESULT
-dif_sram_ctrl_result_t dif_sram_ctrl_exec_lock(
-    const dif_sram_ctrl_t *sram_ctrl);
+dif_sram_ctrl_result_t dif_sram_ctrl_scramble_is_busy(
+    const dif_sram_ctrl_t *sram_ctrl, bool *is_busy);
 
 /**
- * Reports whether the execution enable register is locked or not.
+ * Requests a new scrambling key.
+ *
+ * On successful completion SRAM is effectively scrambled (due to different
+ * logical mapping of the physical addresses). However, it is recommended to
+ * additionally overwrite SRAM with pseudo-random data by calling
+ * `dif_sram_ctrl_wipe`.
```

Comment:
From a security perspective I think scrambling is enough. I think that INIT is needed to avoid parity errors when loading values after scrambling (see https://github.com/lowRISC/opentitan/issues/4731).

Reply:
[LFSR Initialization Feature section](https://docs.opentitan.org/hw/ip/sram_ctrl/doc/index.html):
_"Since the scrambling device uses a block cipher in CTR mode, it is undesirable to initialize the memory with all-zeros from a security perspective, as that would reveal the XOR keystream. To this end, the sram_ctrl contains an LFSR-based initialization mechanism that overwrites the the entire memory with pseudorandom data."_

The recommendation I have extrapolated from the above paragraph from the SRAM Controller documentation. I guess in practice SRAM on power-up should be more or less random and not zeroed, however this was my attempt to not rely on any assumptions.

I can remove the recommendation if you think this is unnecessary?

Reply:
Thinking about it I think the recommendation is sound. If the physical SRAM is known to be zero'd then that presumably would reveal the XOR keystream in the same way as if it were initialized to zero after scrambling, so recommending that the SRAM be initialized to pseudo-random values seems sensible.

(BTW it might be worth mentioning in the comment that the data is also scrambled, not just the addresses.)

Reply:
Done.

> (BTW it might be worth mentioning in the comment that the data is also scrambled, not just the addresses.)

Done.